### ISSUE TITLE
Less results than real Google

### ISSUE THREAD

**Issue by WAZAAAAA0** [hnhx/librex#225]
*(cross-posted in https://github.com/searxng/searxng/issues/2438, https://github.com/benbusby/whoogle-search/issues/1004)*

It's been years. Google-searching through a "privacy search engine frontend" will rarely find as many results as the real Google.

Here's a simple test to verify that: come up with a unique Google query that will find as few results as possible, preferably not in English. For example in my test I used `"sfendazi"` but you might need your own unique query since the results come and go. Perform the same search on every public instance, and observe how many find the same results (if the results contain garbage unrelated stuff, consider it a failure). This was the outcome yesterday as of 2023-05-15:

<details>
  <summary><a href="https://github.com/hnhx/librex#instances">LibreX instances:</a> 20 tested, 0 work</summary>

lmao
</details>
<details>
  <summary><a href="https://github.com/benbusby/whoogle-search#public-instances">Whoogle instances:</a> 17 tested, 3 work</summary>

https://s.tokhmi.xyz
https://whoogle.dcs0.hu
https://whoogle.privacydev.net
</details>
<details>
  <summary><a href="https://searx.space/">SearX/SearXNG instances:</a> 92 tested, 19 work if you tweak a setting, only 1 works with defaults</summary>

(the only one that works with defaults is https://opnxng.com)
https://priv.au
https://xo.wtf
https://offtheradar.info
https://searx.oakleycord.dev
https://searx.cthd.icu
https://ooglester.com
https://search.bus-hit.me
https://myprivatesrx.us
https://coppedge.info
https://search.neet.works
https://search.zzls.xyz
https://search.us.projectsegfau.lt
https://s.frlt.one
https://searx.sev.monster
https://stalk.antelope.day
https://searx.esmailelbob.xyz
https://search.serginho.dev
https://search.cronobox.one
https://searx.mxchange.org
</details>

Those 19 instances I listed think they're "smart" and have set their `Search language` to `[auto]`, which auto-selects it based on your browser headers... or they're simply set to something arbitrary, like `[en-US]`. Choosing `[all]` fixes the problem for them.
Meanwhile, the rest of the instances somehow will not find the correct results even when set to `[all]`. From what I've tested with a local SearXNG instance, adding search query parameter `nfpr=1` (along with the pre-existing `safe=off` and `filter=0`) to `searxng/searx/engines/google.py` fixed it. Here's what they do:
- `nfpr=1` -> Showing results for XXX Search instead for YYY ON
- `safe=off` -> SafeSearch OFF
- `filter=0` -> Include omitted results ON

Changing the `Interface language` is fine. Actually, I'd argue language auto-detection should happen to the interface, not to the search results filter, which would be consistent with how major search engines work.

Honestly, just take the `Search language` option away, it does more harm than good. Or at least make `[all]` the default and lock the option behind huge warning signs with flaming skulls that searching will be seriously degraded for everyone if anything other than `[all]` is selected. People don't understand this is the equivalent setting they're touching (taken from Google's official [advanced search page](https://www.google.ca/advanced_search?hl=en&authuser=0)):
![](https://i.imgur.com/qmrKqwO.png)

# TL;DR
Here's a picture to sum up the problem most search frontends are facing:
![](https://i.imgur.com/F6HWkDR.png)
Proposed fixes:
1. remove `Search language` and default it to `[all]`
2. give `[auto]` to the `Interface language` instead
3. add these 3 parameters to unlock all the possible Google results `?safe=off&nfpr=1&filter=0`

**Comment by megamster** [comment]
Your first suggestion would make it unusable for a lot of queries in most languages

**Comment by WAZAAAAA0** [comment]
> Your first suggestion would make it unusable for a lot of queries in most languages

I'll quote what has been said already on a [SearX issue](https://github.com/searx/searx/issues/1236#issuecomment-822279284) about it:

> After testing a bit with whoogle, you are right the "default (none)" setting is much better than having to select a language. More over this doesn't require to select a language because Google will automatically recognize the language based on the entered words and display the results in the correct language. So much more convenient than having to force a language in the Searx settings every time you make searches in multiple languages.

**Comment by megamster** [comment]
@WAZAAAAA0 that quote of yours actually explains the problem. There are an awful lot of words that different languages have in common and that can have widely different meanings. I've already mentioned the word "coche" in another issue here, which in Portuguese means "carriage" (the old school horse drawn ones) while in Spanish means "car". To add another random example, "pasta" in Portuguese would be what in English one calls "folder".

In addition to the word input, google also seems to use location, IP probably, something librex or whoogle are unable to do, hence the need for that language selector.

I have another open issue here precisely because for image search librex doesn't seem to respect the language selection, making it unusable for a lot of searches

**Comment by WAZAAAAA0** [comment]
I'm aware words are shared in many languages and that if you search, for example, the word `rape` from an Italian IP/instance you'll probably get the vegetable, in French the cheese grater, and in English the assault.
My personal solution to that is to simply do better searches with more context like "rape vegetale".

Normally I'd be all for giving more choice to the user (given that the default should **ALWAYS** be auto) but history has shown us it's not happening. Some instances block access to settings.php entirely. Some reset back to `en` when you try to set to language to empty. Some hide the "Google settings" section. And  the rest, even when set to empty, still don't match the number of real Google results,

I'm tired boss. I just want to proxy my Google searches without being crippled.

**Comment by megamster** [comment]
> My personal solution to that is to simply do better searches with more context like "rape vegetale".

Well, that doesn't always work either, but even if it does it normally has the first 5 or so results correct and the rest are just garbage. You yourself set the criteria:

> if the results contain garbage unrelated stuff, consider it a failure

And then we have the same differences as to the meaning of words within the same language that even with longer queries are impossible to avoid. If anything, librex is missing the ability to set say pt-pt or pt-br like searx has instead of just pt. The same word in Portugal and Brasil can mean very different things. For instance you can be searching for something harmless and get pretty spicy stuff and adding context sometimes just makes it show spicier stuff... Yeah, that happened to me totally at random

**Comment by WAZAAAAA0** [comment]
I see where you're coming from but that's not the "type of garbage" I meant. The garbage most likely got triggered by this Google message

> Showing results for AAA
> Search instead for BBB

which occurs more often when you narrow your search to only 1 language afaik. You're getting the results from AAA (aka GARBAGE, wildly unrelated) rather than the query BBB you were searching for

**Comment by megamster** [comment]
> which occurs more often when you narrow your search to only 1 language afaik. You're getting the results from AAA (aka GARBAGE, wildly unrelated) rather than the query BBB you were searching for

Not really, no. Happens with librex image search, where language settings do not apply. Searxng set to pt-pt has no such problem

> your absolute best fix would be to find a public instance hosted in your country, set to find results in all languages, but good luck with it

Precisely because thats not always possible, language/region settings are needed 😉

**Comment by WAZAAAAA0** [comment]
if an user wants to willingly exclude the vast majority of results AND trusts Google's language filter accuracy as well that's fine by me, then we go back to the initial, less radical suggestion of
> at least make `[all]` the default and lock the option behind huge warning signs with flaming skulls that searching will be seriously degraded for everyone if anything other than `[all]` is selected

**Comment by megamster** [comment]
> if an user wants to willingly exclude the vast majority of results AND trusts Google's language filter accuracy as well that's fine by me

By all means, [all] should be the default, I'm 100% on board with that, though the option to narrrow it down by language should be there and easily available (in the same manner searxng handles it, for example) and I don't see the need for any huge warnings. Just clear labeling, similar to google's: "search only in [language]". It should be obvious that if you're filtering it by language you'll be getting less results. I don't think the auto option serves much of a purpose

**Comment by kubo6472** [comment]
I'm sorry for necroposting a bit. But I'm tired of always switching the language in the search setting between the two that give me correct results: first one being English, second one my native Slovak. Is this fixable the same way in LibreX? (e.g. put "all" in the results language?) Or can I do arrays, like [en,sk]?

**Comment by entrider** [comment]
> In addition to the word input, google also seems to use location, IP probably, something librex or whoogle are unable to do, hence the need for that language selector.

But Google is able to. As a result you shouldnt expect to get the exact same results between instances hosted in different countries (and I believe between all instances in general).

> Is this fixable the same way in LibreX? (e.g. put "all" in the results language?) Or can I do arrays, like [en,sk]?

It's technically possible to make 2 search queries and combine the results of each.

**Comment by kubo6472** [comment]
> It's technically possible to make 2 search queries and combine the results of each.

Using "all" now. Getting a lot of russian results at the top. Putting a language code in the search itself now, helps much more.

**Comment by megamster** [comment]
> As a result you shouldnt expect to get the exact same results between instances hosted in different countries (and I believe between all instances in general).

I dont, but I do expect to be able to configure it in a manner that gives me usable results and not completely random nonsense.. Its precisely the fact that librex instances can't use the same location tools google can that makes the language selector needed, that's the entire point I was making

**Comment by entrider** [comment]
[LibreY ](https://github.com/Ahwxorg/librey/) has it, but only in the settings.
